File: frontend/lib/cron.ts
/**
 * Five-field cron for schedule triggers: the grammar the server accepts
 * (minute hour day-of-month month day-of-week, lists, ranges and steps,
 * digits only, plus the `@hourly` … `@yearly` aliases), a reading of it in
 * words, and the handful of presets a builder offers before it falls back
 * to the raw expression. Timezones are IANA names, checked with the
 * browser's own tables. Nothing here touches the network or React.
 */

export const CRON_ALIASES: Record<string, string> = {
   '@hourly': '0 * * * *',
   '@daily': '0 0 * * *',
   '@weekly': '0 0 * * 0',
   '@monthly': '0 0 1 * *',
   '@yearly': '0 0 1 1 *',
};

const FIELDS = [
   { name: 'minute', min: 0, max: 59 },
   { name: 'hour', min: 0, max: 23 },
   { name: 'day of month', min: 1, max: 31 },
   { name: 'month', min: 1, max: 12 },
   { name: 'day of week', min: 0, max: 7 },
] as const;

const FIELD_PATTERN = /^[0-9*,/-]+$/;

export interface CronField {
   raw: string;
   /** True when the field is `*` alone: unrestricted. */
   star: boolean;
   /** N for a field written as a star over N (every N minutes, hours…), or null. */
   everyStep: number | null;
   /** Every value the field admits, ascending. Day 7 is folded into 0. */
   values: number[];
}

export interface ParsedCron {
   /** The expression with aliases expanded and whitespace normalised. */
   expression: string;
   fields: [CronField, CronField, CronField, CronField, CronField];
}

function parseField(
   raw: string,
   spec: (typeof FIELDS)[number]
): { field: CronField } | { error: string } {
   if (!FIELD_PATTERN.test(raw)) {
      return { error: `The ${spec.name} field allows digits, "*", ",", "-" and "/" only.` };
   }
   const values = new Set<number>();
   let star = false;
   let everyStep: number | null = null;
   for (const item of raw.split(',')) {
      if (item === '') return { error: `The ${spec.name} field has an empty entry.` };
      const [range, stepText, extra] = item.split('/');
      if (extra !== undefined) return { error: `The ${spec.name} field has a stray "/".` };
      let step = 1;
      if (stepText !== undefined) {
         if (!/^\d+$/.test(stepText) || Number(stepText) === 0) {
            return { error: `A step in the ${spec.name} field is a positive number.` };
         }
         step = Number(stepText);
      }
      let low: number;
      let high: number;
      if (range === '*') {
         low = spec.min;
         high = spec.max;
         if (stepText === undefined) star = true;
         else if (raw === item) everyStep = step;
      } else {
         const [lowText, highText, more] = range.split('-');
         if (
            more !== undefined ||
            !/^\d+$/.test(lowText) ||
            (highText !== undefined && !/^\d+$/.test(highText))
         ) {
            return { error: `The ${spec.name} field has a value that is not a number or range.` };
         }
         low = Number(lowText);
         high =
            highText === undefined ? (stepText === undefined ? low : spec.max) : Number(highText);
         if (low < spec.min || low > spec.max || high < spec.min || high > spec.max) {
            return { error: `The ${spec.name} field is between ${spec.min} and ${spec.max}.` };
         }
         if (high < low) return { error: `A range in the ${spec.name} field runs backwards.` };
      }
      for (let value = low; value <= high; value += step) {
         values.add(spec.name === 'day of week' && value === 7 ? 0 : value);
      }
   }
   const list = Array.from(values).sort((a, b) => a - b);
   return { field: { raw, star: star && raw === '*', everyStep, values: list } };
}

/** The expression parsed, or the first thing wrong with it. */
export function parseCron(
   text: string
): { ok: true; cron: ParsedCron } | { ok: false; error: string } {
   let trimmed = text.trim();
   if (trimmed === '') return { ok: false, error: 'Give a schedule.' };
   if (trimmed.startsWith('@')) {
      const alias = CRON_ALIASES[trimmed];
      if (!alias) {
         return { ok: false, error: 'Aliases are @hourly, @daily, @weekly, @monthly and @yearly.' };
      }
      trimmed = alias;
   }
   const parts = trimmed.split(/\s+/);
   if (parts.length !== 5) {
      return {
         ok: false,
         error: 'A cron expression has five fields: minute hour day-of-month month day-of-week.',
      };
   }
   const fields: CronField[] = [];
   for (const [index, part] of parts.entries()) {
      const result = parseField(part, FIELDS[index]);
      if ('error' in result) return { ok: false, error: result.error };
      fields.push(result.field);
   }
   return {
      ok: true,
      cron: {
         expression: parts.join(' '),
         fields: fields as ParsedCron['fields'],
      },
   };
}

/** What is wrong with the expression, or null when the server would accept it. */
export function cronProblem(text: string): string | null {
   const parsed = parseCron(text);
   return parsed.ok ? null : parsed.error;
}

const WEEKDAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
const MONTHS = [
   'January',
   'February',
   'March',
   'April',
   'May',
   'June',
   'July',
   'August',
   'September',
   'October',
   'November',
   'December',
];

export const WEEKDAY_OPTIONS = WEEKDAYS.map((label, value) => ({ value, label }));

function pad(value: number): string {
   return value.toString().padStart(2, '0');
}

function clock(hour: number, minute: number): string {
   return `${pad(hour)}:${pad(minute)}`;
}

function ordinal(day: number): string {
   const rest = day % 100;
   if (rest >= 11 && rest <= 13) return `${day}th`;
   switch (day % 10) {
      case 1:
         return `${day}st`;
      case 2:
         return `${day}nd`;
      case 3:
         return `${day}rd`;
      default:
         return `${day}th`;
   }
}

function joinNames(names: string[]): string {
   if (names.length <= 1) return names.join('');
   return `${names.slice(0, -1).join(', ')} and ${names[names.length - 1]}`;
}

function single(field: CronField): number | null {
   return !field.star && field.values.length === 1 ? field.values[0] : null;
}

const WEEKDAY_VALUES = [1, 2, 3, 4, 5];
const WEEKEND_VALUES = [0, 6];

function sameValues(left: number[], right: number[]): boolean {
   return left.length === right.length && left.every((value, index) => value === right[index]);
}

/**
 * "Every day at 09:00", "Every weekday at 17:30", "Every Monday at 09:00",
 * "Every hour", "Every 15 minutes", "On the 1st of every month at 08:00";
 * anything the words do not cover is shown as the expression itself.
 */
export function describeCron(text: string, timezone?: string | null): string {
   const parsed = parseCron(text);
   const suffix = timezone ? ` · ${timezone}` : '';
   if (!parsed.ok) return `${text.trim() || 'no schedule'}${suffix}`;
   const [minute, hour, day, month, weekday] = parsed.cron.fields;
   const minuteAt = single(minute);
   const hourAt = single(hour);
   let words: string | null = null;
   if (minute.star && hour.star && day.star && month.star && weekday.star) {
      words = 'Every minute';
   } else if (minute.everyStep && hour.star && day.star && month.star && weekday.star) {
      words = `Every ${minute.everyStep} minutes`;
   } else if (minuteAt !== null && hour.star && day.star && month.star && weekday.star) {
      words = minuteAt === 0 ? 'Every hour' : `Every hour at :${pad(minuteAt)}`;
   } else if (minuteAt !== null && hour.everyStep && day.star && month.star && weekday.star) {
      words = `Every ${hour.everyStep} hours${minuteAt === 0 ? '' : ` at :${pad(minuteAt)}`}`;
   } else if (minuteAt !== null && hourAt !== null) {
      const time = clock(hourAt, minuteAt);
      const dayAt = single(day);
      const monthAt = single(month);
      if (day.star && month.star) {
         if (weekday.star) words = `Every day at ${time}`;
         else if (sameValues(weekday.values, WEEKDAY_VALUES)) words = `Every weekday at ${time}`;
         else if (sameValues(weekday.values, WEEKEND_VALUES))
            words = `Every weekend day at ${time}`;
         else
            words = `Every ${joinNames(weekday.values.map((value) => WEEKDAYS[value]))} at ${time}`;
      } else if (dayAt !== null && weekday.star) {
         if (month.star) words = `On the ${ordinal(dayAt)} of every month at ${time}`;
         else if (monthAt !== null)
            words = `Every year on ${dayAt} ${MONTHS[monthAt - 1]} at ${time}`;
      }
   }
   return `${words ?? `cron ${parsed.cron.expression}`}${suffix}`;
}

// ---------------------------------------------------------------------------
// The builder's presets

export type SchedulePreset = 'every_minute' | 'hourly' | 'daily' | 'weekdays' | 'weekly' | 'custom';

export const SCHEDULE_PRESETS: { value: SchedulePreset; label: string; hint: string }[] = [
   { value: 'hourly', label: 'Every hour', hint: 'At a minute past each hour' },
   { value: 'daily', label: 'Every day at', hint: 'One time a day' },
   { value: 'weekdays', label: 'Every weekday at', hint: 'Monday to Friday' },
   { value: 'weekly', label: 'Every week on', hint: 'One day a week' },
   { value: 'every_minute', label: 'Every minute', hint: 'For trying a workflow out' },
   { value: 'custom', label: 'Custom cron', hint: 'Five fields, as on a server' },
];

export interface ScheduleDraft {
   preset: SchedulePreset;
   hour: number;
   minute: number;
   weekday: number;
   /** The expression as typed for the custom preset; derived for the others. */
   cron: string;
}

export function buildCron(draft: ScheduleDraft): string {
   switch (draft.preset) {
      case 'every_minute':
         return '* * * * *';
      case 'hourly':
         return `${draft.minute} * * * *`;
      case 'daily':
         return `${draft.minute} ${draft.hour} * * *`;
      case 'weekdays':
         return `${draft.minute} ${draft.hour} * * 1-5`;
      case 'weekly':
         return `${draft.minute} ${draft.hour} * * ${draft.weekday}`;
      case 'custom':
         return draft.cron;
   }
}

/** The preset an expression came from, so a stored schedule opens on its own controls. */
export function scheduleDraftFromCron(text: string): ScheduleDraft {
   const base: ScheduleDraft = { preset: 'custom', hour: 9, minute: 0, weekday: 1, cron: text };
   const parsed = parseCron(text);
   if (!parsed.ok) return base;
   const [minute, hour, day, month, weekday] = parsed.cron.fields;
   const minuteAt = single(minute);
   const hourAt = single(hour);
   if (!day.star || !month.star) return base;
   if (minute.star && hour.star && weekday.star) return { ...base, preset: 'every_minute' };
   if (minuteAt === null) return base;
   if (hour.star && weekday.star) return { ...base, preset: 'hourly', minute: minuteAt };
   if (hourAt === null) return base;
   if (weekday.star) return { ...base, preset: 'daily', hour: hourAt, minute: minuteAt };
   if (sameValues(weekday.values, WEEKDAY_VALUES)) {
      return { ...base, preset: 'weekdays', hour: hourAt, minute: minuteAt };
   }
   const weekdayAt = single(weekday);
   if (weekdayAt !== null) {
      return { ...base, preset: 'weekly', hour: hourAt, minute: minuteAt, weekday: weekdayAt };
   }
   return base;
}

// ---------------------------------------------------------------------------
// Timezones

const FALLBACK_TIMEZONES = [
   'UTC',
   'Europe/London',
   'Europe/Paris',
   'Europe/Berlin',
   'Europe/Rome',
   'Europe/Madrid',
   'America/New_York',
   'America/Chicago',
   'America/Denver',
   'America/Los_Angeles',
   'America/Sao_Paulo',
   'Asia/Tokyo',
   'Asia/Shanghai',
   'Asia/Kolkata',
   'Asia/Singapore',
   'Australia/Sydney',
];

/** The browser's own timezone, or UTC when it will not say. */
export function browserTimezone(): string {
   try {
      return Intl.DateTimeFormat().resolvedOptions().timeZone || 'UTC';
   } catch {
      return 'UTC';
   }
}

let cachedTimezones: string[] | null = null;

/** Every IANA name the browser knows, the browser's own included, sorted. */
export function timezoneOptions(): string[] {
   if (cachedTimezones) return cachedTimezones;
   let names: string[] = FALLBACK_TIMEZONES;
   try {
      const supported = Intl.supportedValuesOf('timeZone');
      if (supported.length > 0) names = supported;
   } catch {
      // An older engine: the short list will do.
   }
   const set = new Set(names);
   set.add('UTC');
   set.add(browserTimezone());
   cachedTimezones = Array.from(set).sort((a, b) => a.localeCompare(b));
   return cachedTimezones;
}

/** True when the browser can format dates in the zone, which is what the server checks too. */
export function isValidTimezone(name: string): boolean {
   if (!name || name === 'Local') return false;
   try {
      new Intl.DateTimeFormat('en-US', { timeZone: name });
      return true;
   } catch {
      return false;
   }
}

/** What is wrong with a schedule trigger's config, or null when the server would accept it. */
export function scheduleProblem(value: {
   cron?: string | null;
   timezone?: string | null;
}): string | null {
   const cron = cronProblem(value.cron ?? '');
   if (cron) return cron;
   if (!isValidTimezone(value.timezone ?? '')) return 'Pick an IANA timezone.';
   return null;
}

/** "26 Aug 10:18" on the zone's wall clock, for a schedule's fire instant. */
export function formatInstant(iso: string, timezone?: string | null): string {
   const date = new Date(iso);
   if (Number.isNaN(date.getTime())) return iso;
   try {
      return new Intl.DateTimeFormat('en-GB', {
         day: 'numeric',
         month: 'short',
         hour: '2-digit',
         minute: '2-digit',
         hour12: false,
         timeZone: timezone && isValidTimezone(timezone) ? timezone : undefined,
      })
         .format(date)
         .replace(',', '');
   } catch {
      return iso;
   }
}

File: server/internal/cron/cron.go
// Package cron parses five-field cron expressions and computes fire times in
// a named timezone.
//
// It exists so schedule triggers (P4.5) never depend on robfig/cron, which is
// only an indirect dependency pulled in by Temporal and whose v1 API is
// unmaintained. The grammar is the classic one: minute, hour, day of month,
// month and day of week, each a list of values, ranges and steps, plus the
// @hourly … @yearly aliases. Names are not accepted: the validator already
// refuses letters, and a numeric-only grammar has no locale.
//
// Fire times are computed on the wall clock of the schedule's timezone and
// then resolved to an instant, so "09:00 Europe/Rome" fires at 08:00 UTC in
// winter and 07:00 UTC in summer. A wall-clock time that does not exist on a
// spring-forward day is skipped; one that exists twice on a fall-back day
// fires once.
package cron

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Schedule is a parsed expression.
type Schedule struct {
	expression string
	minutes    [60]bool
	hours      [24]bool
	days       [32]bool
	months     [13]bool
	weekdays   [7]bool
	// anyDay and anyWeekday record a "*" in the day-of-month and day-of-week
	// fields: when both are restricted a day matches either (the POSIX
	// rule), when one is "*" the other decides alone.
	anyDay     bool
	anyWeekday bool
}

// lookahead bounds the search for the next fire time. An expression that
// never matches ("0 0 31 2 *") stops here instead of spinning forever.
const lookahead = 5

var aliases = map[string]string{
	"@hourly":  "0 * * * *",
	"@daily":   "0 0 * * *",
	"@weekly":  "0 0 * * 0",
	"@monthly": "0 0 1 * *",
	"@yearly":  "0 0 1 1 *",
}

type field struct {
	name string
	min  int
	max  int
}

var fields = [5]field{
	{"minute", 0, 59}, {"hour", 0, 23}, {"day of month", 1, 31}, {"month", 1, 12}, {"day of week", 0, 7},
}

// Parse reads an expression. The error names the field that is wrong and
// never echoes more than the offending token.
func Parse(expression string) (Schedule, error) {
	trimmed := strings.TrimSpace(expression)
	if alias, ok := aliases[trimmed]; ok {
		trimmed = alias
	}
	parts := strings.Fields(trimmed)
	if len(parts) != 5 {
		return Schedule{}, errors.New("cron: an expression has five fields: minute hour day-of-month month day-of-week")
	}
	schedule := Schedule{expression: strings.Join(parts, " ")}
	sets := [5][]bool{
		schedule.minutes[:], schedule.hours[:], schedule.days[:], schedule.months[:], make([]bool, 8),
	}
	for index, part := range parts {
		spec := fields[index]
		star, err := parseField(part, spec, sets[index])
		if err != nil {
			return Schedule{}, err
		}
		switch index {
		case 2:
			schedule.anyDay = star
		case 4:
			schedule.anyWeekday = star
		}
	}
	// 7 is Sunday too; fold it onto 0 so the weekday table has one entry.
	for day := 0; day < 7; day++ {
		schedule.weekdays[day] = sets[4][day]
	}
	if sets[4][7] {
		schedule.weekdays[0] = true
	}
	return schedule, nil
}

// parseField fills set with the values a field names and reports whether the
// field was an unrestricted "*".
func parseField(text string, spec field, set []bool) (bool, error) {
	if text == "" {
		return false, fmt.Errorf("cron: the %s field is empty", spec.name)
	}
	star := false
	for _, item := range strings.Split(text, ",") {
		base, step := item, 1
		if before, after, ok := strings.Cut(item, "/"); ok {
			parsed, err := strconv.Atoi(after)
			if err != nil || parsed < 1 || parsed > spec.max {
				return false, fmt.Errorf("cron: the %s field has an invalid step %q", spec.name, after)
			}
			base, step = before, parsed
		}
		low, high := spec.min, spec.max
		switch {
		case base == "*":
			if step == 1 && item == "*" {
				star = true
			}
		case strings.Contains(base, "-"):
			from, to, _ := strings.Cut(base, "-")
			var err error
			if low, err = parseValue(from, spec); err != nil {
				return false, err
			}
			if high, err = parseValue(to, spec); err != nil {
				return false, err
			}
			if low > high {
				return false, fmt.Errorf("cron: the %s field has a reversed range %q", spec.name, base)
			}
		default:
			value, err := parseValue(base, spec)
			if err != nil {
				return false, err
			}
			low = value
			if step == 1 {
				high = value
			}
		}
		for value := low; value <= high; value += step {
			set[value] = true
		}
	}
	return star, nil
}

func parseValue(text string, spec field) (int, error) {
	value, err := strconv.Atoi(text)
	if err != nil || value < spec.min || value > spec.max {
		return 0, fmt.Errorf("cron: the %s field has an invalid value %q (%d..%d)", spec.name, text, spec.min, spec.max)
	}
	return value, nil
}

// String returns the normalised five-field expression.
func (schedule Schedule) String() string {
	return schedule.expression
}

// wall is a wall-clock instant without a zone.
type wall struct {
	year, month, day, hour, minute int
}

func (clock wall) weekday() time.Weekday {
	return time.Date(clock.year, time.Month(clock.month), clock.day, 0, 0, 0, 0, time.UTC).Weekday()
}

// normalise rolls overflowed fields the way the calendar does.
func (clock wall) normalise() wall {
	rolled := time.Date(clock.year, time.Month(clock.month), clock.day, clock.hour, clock.minute, 0, 0, time.UTC)
	return wall{rolled.Year(), int(rolled.Month()), rolled.Day(), rolled.Hour(), rolled.Minute()}
}

func (schedule Schedule) matchesDay(clock wall) bool {
	day := schedule.days[clock.day]
	weekday := schedule.weekdays[clock.weekday()]
	switch {
	case schedule.anyDay && schedule.anyWeekday:
		return true
	case schedule.anyDay:
		return weekday
	case schedule.anyWeekday:
		return day
	default:
		return day || weekday
	}
}

// Next returns the first fire time strictly after the instant, in the
// location, or false when none exists within the lookahead.
func (schedule Schedule) Next(after time.Time, location *time.Location) (time.Time, bool) {
	if location == nil {
		location = time.UTC
	}
	local := after.In(location)
	clock := wall{local.Year(), int(local.Month()), local.Day(), local.Hour(), local.Minute() + 1}.normalise()
	limit := local.Year() + lookahead
	for clock.year <= limit {
		switch {
		case !schedule.months[clock.month]:
			clock = wall{clock.year, clock.month + 1, 1, 0, 0}.normalise()
		case !schedule.matchesDay(clock):
			clock = wall{clock.year, clock.month, clock.day + 1, 0, 0}.normalise()
		case !schedule.hours[clock.hour]:
			clock = wall{clock.year, clock.month, clock.day, clock.hour + 1, 0}.normalise()
		case !schedule.minutes[clock.minute]:
			clock = wall{clock.year, clock.month, clock.day, clock.hour, clock.minute + 1}.normalise()
		default:
			candidate := time.Date(clock.year, time.Month(clock.month), clock.day, clock.hour, clock.minute, 0, 0, location)
			// A wall-clock minute the zone skipped resolves to another
			// minute; it never existed, so it never fires.
			if candidate.Day() != clock.day || candidate.Hour() != clock.hour || candidate.Minute() != clock.minute || !candidate.After(after) {
				clock = wall{clock.year, clock.month, clock.day, clock.hour, clock.minute + 1}.normalise()
				continue
			}
			return candidate, true
		}
	}
	return time.Time{}, false
}

// NextIn is Next with the location named as an IANA zone.
func (schedule Schedule) NextIn(after time.Time, timezone string) (time.Time, error) {
	location, err := time.LoadLocation(timezone)
	if err != nil || timezone == "" || timezone == "Local" {
		return time.Time{}, fmt.Errorf("cron: %q is not an IANA timezone", timezone)
	}
	next, ok := schedule.Next(after, location)
	if !ok {
		return time.Time{}, errors.New("cron: the expression never fires within the next five years")
	}
	return next, nil
}
